### PR TITLE
New: Add no-unreachable-loop rule (fixes #12381)

### LINKED ISSUE #12381: New rule proposal: no-unreachable-loop
**Please describe what the rule should do:**

Disallows loops which body exits the loop in all paths and therefore can never even reach the test condition for the second time (or first time if it's `do-while`), meaning that it isn't actually a loop.

Targets `while`, `do-while`, `for`, `for-in` and `for-of` loops.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
while (foo) {
    doSomething(foo);
    foo = foo.parent;
    break;
}

for (i = 0; i < arr.length; i++) {
    if (isSomething(arr[i])) {
        break;
    } else {
        return false;
    }
}

do {
    if (foo.type === "bar") {
        return foo;
    } else {
        throw new Error("invalid");
    }
} while (foo);

for (foo of bar) {
    if (foo.type === "bar") {
        doSomething(foo);
    }
    break;
}

for (foo in bar) break;
```

The code is correct only if the end of the body is reachable (i.e. loops from the end) or the loop has a `continue` statement.

**Why should this rule be included in ESLint (instead of a plugin)?**

It's a very possible error in the code, something is missing or something else is at the wrong place.

Even if it isn't an error the code should be refactored to avoid unnecessary loop syntax.

This is a similar type of error as `no-unreachable`, but I think this should be a separate rule because there is no particular unreachable code in these cases (except for the code in `do-while` test and `for-loop` update, but I think that `no-unreachable` doesn't report that).

In some cases both `no-unreachable` and this rule can report warnings (e.g. `no-unreachable` for code after the last `break`, this rule for the whole loop), but I think that isn't overlapping.

**Are you willing to submit a pull request to implement this rule?**

Yes, I'd love to work on this.

### PR DESCRIPTION
**Pull request by mdjermanovic**
**What is the purpose of this pull request? (put an "X" next to item)**

[X] New rule 

fixes #12381

Examples of **incorrect** code for this rule:

```js
/*eslint no-unreachable-loop: "error"*/

while (foo) {
    doSomething(foo);
    foo = foo.parent;
    break;
}

function verifyList(head) {
    let item = head;
    do {
        if (verify(item)) {
            return true;
        } else {
            return false;
        }
    } while (item)
}

function findSomething(arr) {
    for (var i = 0; i < arr.length; i++) {
        if (isSomething(arr[i])) {
            return arr[i];
        } else {
            throw new Error("Doesn't exist.");
        }
    }
}

for (key in obj) {
    if (key.startsWith("_")) {
        break;
    }
    firstKey = key;
    firstValue = obj[key];
    break;
}

for (foo of bar) {
    if (foo.id === id) {
        doSomething(foo);
    }
    break;
}
```

**What changes did you make? (Give an overview)**

new rule: `no-unreachable-loop`

**Is there anything you'd like reviewers to focus on?**

### CODE REVIEW

File: tests/lib/rules/no-unreachable-loop.js [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+    "{ return; }",
+    "throw err;",
+    "{ throw err; }",
+    "{ foo(); break; }",
+    "{ break; foo(); }",
+    "if (foo) break; else return;",
+    "{ if (foo) { return; } else { break; } bar(); }",
+    "{ if (foo) { return; } throw err; }",
+    "{ switch (foo) { default: throw err; } }",
+    "{ switch (foo) { case 1: throw err; default: return; } }",
```
Comment (aladdin-add) [inline tests/lib/rules/no-unreachable-loop.js:112]: good to have a case like:
```js
"{ switch (foo) { case 1: something(); default: return; } }",
```

File: tests/lib/rules/no-unreachable-loop.js [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+/**
+ * Creates source code from the given loop template and loop body.
+ * @param {string} template Loop template.
+ * @param {string} body Loop body.
+ * @returns {string} Source code.
+ */
+function getSourceCode(template, body) {
+    const loop = template.replace("<body>", body);
+
+    return body.includes("return") && !template.includes("function") ? `function someFunc() { ${loop} }` : loop;
```
Comment (aladdin-add) [inline tests/lib/rules/no-unreachable-loop.js:152]: just want to point out: `string.include()` is not a reliable way to do the check, e.g.
```js
let s = "return 1;";
// or
arr.map(function (){return 1;});
```
But it seems no harm wrapping these cases in function bodies. 😄

**Comment by nzakas** [comment]
@mdjermanovic are you still working on this?

**Comment by mdjermanovic** [comment]
Yes, sorry I didn't have much time last week,

File: tests/lib/rules/no-unreachable-loop.js (hunk 2 above)
Comment (mdjermanovic) [reply tests/lib/rules/no-unreachable-loop.js:152]: Agreed, this is far from ideal, but couldn't figure out a better way to test all combinations (and not add too much complexity at the same time).

File: tests/lib/rules/no-unreachable-loop.js (hunk 1 above)
Comment (mdjermanovic) [reply tests/lib/rules/no-unreachable-loop.js:112]: Done, thanks!

**Pull request merged**